src/daft-table/src/ops/joins/hash_join.rs

```diff
@@ -1,74 +1,372 @@
+
+        let r_hashes = rkeys.hash_rows()?;
+
+        let is_equal = build_multi_array_is_equal(
+            lkeys.columns.as_slice(),
+            rkeys.columns.as_slice(),
+            false,
+            false,
+        )?;
+
+        let mut left_idx = vec![];
+        let mut right_idx = vec![];
+
+        for (r_idx, h) in r_hashes.as_arrow().values_iter().enumerate() {
+            if let Some((_, indices)) = probe_table.raw_entry().from_hash(*h, |other| {
+                *h == other.hash && {
+                    let l_idx = other.idx;
+                    is_equal(l_idx as usize, r_idx)
+                }
+            }) {
+                for l_idx in indices {
+                    left_idx.push(*l_idx);
+                    right_idx.push(r_idx as u64);
+                }
+            }
+        }
+
+        let lseries = UInt64Array::from(("left_indices", left_idx)).into_series();
+        let rseries = UInt64Array::from(("right_indices", right_idx)).into_series();
+
+        if probe_left {
+            (lseries, rseries)
+        } else {
+            (rseries, lseries)
+        }
+    };
+
+    let mut join_series = left
+        .get_columns(lkeys.column_names().as_slice())?
+        .take(&lidx)?
+        .columns;
+
+    drop(lkeys);
+    drop(rkeys);
+
+    join_series =
+        add_non_join_key_columns(left, right, lidx, ridx, left_on, right_on, join_series)?;
+
+    Table::new(join_schema, join_series)
+}
+
+pub(super) fn hash_left_right_join(
+    left: &Table,
+    right: &Table,
+    left_on: &[ExprRef],
+    right_on: &[ExprRef],
+    left_side: bool,
+) -> DaftResult<Table> {
+    let join_schema = infer_join_schema(&left.schema, &right.schema, left_on, right_on)?;
+    let lkeys = left.eval_expression_list(left_on)?;
+    let rkeys = right.eval_expression_list(right_on)?;
+
+    let (lkeys, rkeys) = match_types_for_tables(&lkeys, &rkeys)?;
+
+    // For left joins, swap tables and use right join, since we want to probe the side we are joining from
+    let (lkeys, rkeys) = if left_side {
+        (rkeys, lkeys)
+    } else {
+        (lkeys, rkeys)
+    };
+
+    let (lidx, ridx) = if lkeys.columns.iter().any(|s| s.data_type().is_null())
+        || rkeys.columns.iter().any(|s| s.data_type().is_null())
+    {
+        (
+            UInt64Array::full_null("left_indices", &DataType::UInt64, rkeys.len()).into_series(),
+            UInt64Array::from((
+                "right_indices",
+                (0..(rkeys.len() as u64)).collect::<Vec<_>>(),
+            ))
+            .into_series(),
+        )
+    } else {
+        let probe_table = lkeys.to_probe_hash_table()?;
+
+        let r_hashes = rkeys.hash_rows()?;
+
+        let is_equal = build_multi_array_is_equal(
+            lkeys.columns.as_slice(),
+            rkeys.columns.as_slice(),
+            false,
+            false,
+        )?;
+
+        let mut left_idx = vec![];
+        let mut right_idx = vec![];
```

Comment:
We know that `right_idx` will be at least the length of `r_idx` so we should reserve the capacity.
```suggestion
        let mut right_idx = Vec::with_capacity(r_idx.len());
```